File: backend-java/src/main/java/com/smartcrop/service/AdvisoryService.java
package com.smartcrop.service;

import com.smartcrop.config.AppProperties;
import com.smartcrop.exception.AppException;
import com.smartcrop.model.SoilProfile;
import com.smartcrop.repository.SoilProfileRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.web.client.ResourceAccessException;
import org.springframework.web.client.RestTemplate;

import java.math.BigDecimal;
import java.util.*;

@Service
@RequiredArgsConstructor
public class AdvisoryService {

    private final SoilProfileRepository soilProfileRepository;
    private final AppProperties appProperties;
    private final RestTemplate restTemplate;

    @SuppressWarnings("unchecked")
    public Object getCropRecommendations(UUID farmerId, UUID plotId) {
        SoilProfile profile = soilProfileRepository.findByIdAndFarmerId(plotId, farmerId)
            .orElseThrow(() -> AppException.notFound("Soil profile not found"));

        if (profile.getPh() == null || profile.getNitrogen() == null
                || profile.getPhosphorus() == null || profile.getPotassium() == null
                || profile.getSoilType() == null) {
            throw new AppException("INCOMPLETE_SOIL_PROFILE",
                "Soil profile is missing required fields. Please complete ph, nitrogen, phosphorus, potassium, and soil type.",
                HttpStatus.BAD_REQUEST);
        }

        // Try external advisory engine first
        try {
            Map<String, Object> body = new HashMap<>();
            Map<String, Object> soilMap = new HashMap<>();
            soilMap.put("type", profile.getSoilType());
            soilMap.put("ph", profile.getPh());
            soilMap.put("n", profile.getNitrogen());
            soilMap.put("p", profile.getPhosphorus());
            soilMap.put("k", profile.getPotassium());
            body.put("soil_profile", soilMap);
            body.put("location", Map.of(
                "lat", profile.getLatitude() != null ? profile.getLatitude() : 0,
                "lon", profile.getLongitude() != null ? profile.getLongitude() : 0
            ));
            body.put("season", "kharif");
            body.put("crop_history", List.of());

            Object result = restTemplate.postForObject(
                appProperties.getAdvisoryEngineUrl() + "/internal/advisory/crops",
                body, Map.class);
            if (result != null) return result;
        } catch (Exception ignored) {
            // Fall through to built-in engine
        }

        // Built-in rule-based crop recommendation
        return builtInCropRecommendations(profile);
    }

    @SuppressWarnings("unchecked")
    public Object getFertilizerGuidance(UUID farmerId, UUID plotId, String cropId) {
        SoilProfile profile = soilProfileRepository.findByIdAndFarmerId(plotId, farmerId)
            .orElseThrow(() -> new AppException("NO_SOIL_PROFILE",
                "No soil profile found. Please create a soil profile first.",
                HttpStatus.BAD_REQUEST));

        try {
            Map<String, Object> body = new HashMap<>();
            Map<String, Object> soilMap = new HashMap<>();
            soilMap.put("type", profile.getSoilType() != null ? profile.getSoilType() : "loamy");
            soilMap.put("ph", profile.getPh() != null ? profile.getPh() : 6.5);
            soilMap.put("n", profile.getNitrogen() != null ? profile.getNitrogen() : 0);
            soilMap.put("p", profile.getPhosphorus() != null ? profile.getPhosphorus() : 0);
            soilMap.put("k", profile.getPotassium() != null ? profile.getPotassium() : 0);
            body.put("soil_profile", soilMap);
            body.put("crop", cropId);

            Object result = restTemplate.postForObject(
                appProperties.getAdvisoryEngineUrl() + "/internal/advisory/fertilizer",
                body, Map.class);
            if (result != null) return result;
        } catch (Exception ignored) {
            // Fall through to built-in engine
        }

        return builtInFertilizerGuidance(profile, cropId);
    }

    // ── Built-in rule-based engine ────────────────────────────────────────────

    private List<Map<String, Object>> builtInCropRecommendations(SoilProfile p) {
        double ph = p.getPh() != null ? p.getPh().doubleValue() : 6.5;
        double n  = p.getNitrogen() != null ? p.getNitrogen().doubleValue() : 0;
        double pp = p.getPhosphorus() != null ? p.getPhosphorus().doubleValue() : 0;
        double k  = p.getPotassium() != null ? p.getPotassium().doubleValue() : 0;
        String soil = p.getSoilType() != null ? p.getSoilType().toLowerCase() : "loamy";

        // Score each crop based on soil parameters
        List<CropScore> scores = new ArrayList<>();

        // Wheat: pH 6-7.5, moderate N, loamy/clay
        scores.add(new CropScore("Wheat",
            score(ph, 6.0, 7.5) + score(n, 40, 120) + soilMatch(soil, "loamy", "clay", "alluvial"),
            "15-25 quintals/acre", "Medium (450-600mm)", "₹8,000-12,000/acre"));

        // Rice: pH 5.5-7, high N, clay/loamy
        scores.add(new CropScore("Rice",
            score(ph, 5.5, 7.0) + score(n, 60, 150) + soilMatch(soil, "clay", "loamy", "alluvial"),
            "18-28 quintals/acre", "High (1200-1800mm)", "₹10,000-15,000/acre"));

        // Maize: pH 5.8-7.5, high N+P, loamy/sandy loam
        scores.add(new CropScore("Maize",
            score(ph, 5.8, 7.5) + score(n, 50, 130) + score(pp, 20, 60) + soilMatch(soil, "loamy", "sandy", "alluvial"),
            "20-30 quintals/acre", "Medium (500-700mm)", "₹7,000-11,000/acre"));

        // Soybean: pH 6-7, moderate N+P+K, loamy
        scores.add(new CropScore("Soybean",
            score(ph, 6.0, 7.0) + score(n, 30, 80) + score(pp, 15, 50) + soilMatch(soil, "loamy", "clay"),
            "8-14 quintals/acre", "Medium (450-700mm)", "₹6,000-9,000/acre"));

        // Cotton: pH 6-8, high K, black/clay
        scores.add(new CropScore("Cotton",
            score(ph, 6.0, 8.0) + score(k, 40, 120) + soilMatch(soil, "black", "clay", "loamy"),
            "6-12 quintals/acre", "Medium (500-800mm)", "₹12,000-18,000/acre"));

        // Sugarcane: pH 6-7.5, high N+K, loamy/alluvial
        scores.add(new CropScore("Sugarcane",
            score(ph, 6.0, 7.5) + score(n, 80, 200) + score(k, 50, 150) + soilMatch(soil, "loamy", "alluvial", "clay"),
            "250-400 quintals/acre", "High (1500-2500mm)", "₹20,000-30,000/acre"));

        // Groundnut: pH 6-7, moderate P+K, sandy loam
        scores.add(new CropScore("Groundnut",
            score(ph, 6.0, 7.0) + score(pp, 20, 60) + score(k, 30, 80) + soilMatch(soil, "sandy", "loamy"),
            "6-10 quintals/acre", "Low-Medium (400-600mm)", "₹8,000-12,000/acre"));

        // Tomato: pH 6-7, high N+P, loamy
        scores.add(new CropScore("Tomato",
            score(ph, 6.0, 7.0) + score(n, 60, 120) + score(pp, 30, 80) + soilMatch(soil, "loamy", "sandy"),
            "80-150 quintals/acre", "Medium (600-1200mm)", "₹15,000-25,000/acre"));

        // Onion: pH 6-7.5, moderate N+K, loamy
        scores.add(new CropScore("Onion",
            score(ph, 6.0, 7.5) + score(n, 40, 100) + score(k, 40, 100) + soilMatch(soil, "loamy", "sandy"),
            "80-120 quintals/acre", "Medium (350-550mm)", "₹12,000-18,000/acre"));

        // Mustard: pH 6-7.5, moderate N, loamy/sandy
        scores.add(new CropScore("Mustard",
            score(ph, 6.0, 7.5) + score(n, 30, 80) + soilMatch(soil, "loamy", "sandy", "alluvial"),
            "6-10 quintals/acre", "Low (250-400mm)", "₹5,000-8,000/acre"));

        scores.sort((a, b) -> Double.compare(b.score, a.score));

        List<Map<String, Object>> result = new ArrayList<>();
        for (int i = 0; i < Math.min(5, scores.size()); i++) {
            CropScore cs = scores.get(i);
            Map<String, Object> m = new LinkedHashMap<>();
            m.put("rank", i + 1);
            m.put("cropName", cs.name);
            m.put("expectedYieldRange", cs.yield);
            m.put("waterRequirement", cs.water);
            m.put("estimatedInputCost", cs.cost);
            m.put("suitabilityScore", Math.round(cs.score * 10.0) / 10.0);
            result.add(m);
        }
        return result;
    }

    private Map<String, Object> builtInFertilizerGuidance(SoilProfile p, String cropId) {
        double n = p.getNitrogen() != null ? p.getNitrogen().doubleValue() : 0;
        double pp = p.getPhosphorus() != null ? p.getPhosphorus().doubleValue() : 0;
        double k = p.getPotassium() != null ? p.getPotassium().doubleValue() : 0;

        // Recommended levels for most crops
        double recN = 80, recP = 40, recK = 40;

        double urea = Math.max(0, (recN - n) * 2.17);   // 1kg N = 2.17kg Urea
        double dap  = Math.max(0, (recP - pp) * 5.43);  // 1kg P = 5.43kg DAP
        double mop  = Math.max(0, (recK - k) * 1.67);   // 1kg K = 1.67kg MOP

        Map<String, Object> result = new LinkedHashMap<>();
        result.put("crop", cropId);
        result.put("recommendations", List.of(
            Map.of("fertilizer", "Urea (46% N)", "quantity", String.format("%.1f kg/acre", urea),
                   "timing", "Split: 50% at sowing, 25% at 30 days, 25% at 60 days"),
            Map.of("fertilizer", "DAP (18-46-0)", "quantity", String.format("%.1f kg/acre", dap),
                   "timing", "Full dose at sowing as basal"),
            Map.of("fertilizer", "MOP (0-0-60)", "quantity", String.format("%.1f kg/acre", mop),
                   "timing", "Full dose at sowing as basal")
        ));
        result.put("soilHealth", Map.of(
            "nitrogenStatus", n < 40 ? "Low — needs supplementation" : n < 80 ? "Medium" : "Adequate",
            "phosphorusStatus", pp < 20 ? "Low — needs supplementation" : pp < 40 ? "Medium" : "Adequate",
            "potassiumStatus", k < 20 ? "Low — needs supplementation" : k < 40 ? "Medium" : "Adequate"
        ));
        return result;
    }

    private double score(double val, double min, double max) {
        if (val >= min && val <= max) return 3.0;
        double dist = val < min ? min - val : val - max;
        return Math.max(0, 3.0 - dist / 20.0);
    }

    private double soilMatch(String soil, String... types) {
        for (String t : types) {
            if (soil.contains(t)) return 2.0;
        }
        return 0.5;
    }

    private static class CropScore {
        String name, yield, water, cost;
        double score;
        CropScore(String name, double score, String yield, String water, String cost) {
            this.name = name; this.score = score;
            this.yield = yield; this.water = water; this.cost = cost;
        }
    }
}

File: backend/src/services/advisoryService.ts
import axios from 'axios';
import { config } from '../config';
import { query } from '../db';
import { AppError } from './userService';
import { getSoilProfile } from './soilProfileService';

// ─── Types ────────────────────────────────────────────────────────────────────

export interface CropRecommendation {
  name: string;
  yieldRange: { min: number; max: number };
  waterRequirement: string;
  estimatedInputCost: number;
}

export interface FertilizerSchedule {
  schedule: Array<{ type: string; quantity: number; unit: string; timing: string }>;
  organicAlternatives: Array<{ type: string; quantity: number; unit: string; timing: string }>;
  soilAmendments: Array<{ type: string; quantity: number; unit: string; reason: string }>;
}

interface CropHistoryRow {
  crop_name: string;
  season: string | null;
  year: number | null;
}

// ─── Service functions ────────────────────────────────────────────────────────

/**
 * Get crop recommendations for a farmer's plot.
 * Fetches soil profile and crop history, then calls the Advisory Engine.
 * Throws INCOMPLETE_SOIL_PROFILE if required soil fields are missing.
 * Throws ADVISORY_ENGINE_UNAVAILABLE if the engine is unreachable or returns 503.
 */
export async function getCropRecommendations(
  farmerId: string,
  plotId: string
): Promise<CropRecommendation[]> {
  const soilProfile = await getSoilProfile(plotId, farmerId);

  // Validate required fields
  if (
    soilProfile.ph == null ||
    soilProfile.nitrogen == null ||
    soilProfile.phosphorus == null ||
    soilProfile.potassium == null ||
    soilProfile.soilType == null
  ) {
    throw new AppError(
      'INCOMPLETE_SOIL_PROFILE',
      'Soil profile is missing required fields. Please complete ph, nitrogen, phosphorus, potassium, and soil type.'
    );
  }

  // Fetch crop history for this plot
  let cropHistory: string[] = [];
  try {
    const historyResult = await query<CropHistoryRow>(
      `SELECT crop_name, season, year
       FROM crop_history
       WHERE farmer_id = $1 AND soil_profile_id = $2
       ORDER BY year DESC, created_at DESC`,
      [farmerId, plotId]
    );
    cropHistory = historyResult.rows.map((r) => r.crop_name);
  } catch (dbErr) {
    // Database not available, proceed with empty crop history
    cropHistory = [];
  }

  try {
    const response = await axios.post<{ crops: any[] }>(
      `${config.advisoryEngineUrl}/internal/advisory/crops`,
      {
        soil_profile: {
          type: soilProfile.soilType,
          ph: soilProfile.ph,
          n: soilProfile.nitrogen,
          p: soilProfile.phosphorus,
          k: soilProfile.potassium,
        },
        location: {
          lat: soilProfile.latitude ?? null,
          lon: soilProfile.longitude ?? null,
        },
        season: getCurrentSeason(),
        crop_history: cropHistory,
      }
    );

    return response.data.crops.map((crop: any) => ({
      name: crop.name,
      yieldRange: {
        min: crop.yield_range?.min ?? crop.yieldRange?.min ?? 0,
        max: crop.yield_range?.max ?? crop.yieldRange?.max ?? 0,
      },
      waterRequirement: crop.water_requirement ?? crop.waterRequirement ?? 'medium',
      estimatedInputCost: crop.estimated_input_cost ?? crop.estimatedInputCost ?? 0,
    }));
  } catch (err: unknown) {
    if (isAdvisoryEngineUnavailable(err)) {
      // Return mock data when Advisory Engine is unavailable
      return getMockCropRecommendations(soilProfile.ph, soilProfile.soilType, cropHistory);
    }
    throw err;
  }
}

/**
 * Get fertilizer guidance for a farmer's plot and selected crop.
 * Throws NO_SOIL_PROFILE if no soil profile is found.
 * Throws ADVISORY_ENGINE_UNAVAILABLE if the engine is unreachable or returns 503.
 */
export async function getFertilizerGuidance(
  farmerId: string,
  plotId: string,
  cropId: string
): Promise<FertilizerSchedule> {
  let soilProfile;
  try {
    soilProfile = await getSoilProfile(plotId, farmerId);
  } catch (err: unknown) {
    if (err instanceof AppError && err.code === 'NOT_FOUND') {
      throw new AppError('NO_SOIL_PROFILE', 'No soil profile found. Please create a soil profile first.');
    }
    throw err;
  }

  try {
    const response = await axios.post<any>(
      `${config.advisoryEngineUrl}/internal/advisory/fertilizer`,
      {
        soil_profile: {
          type: soilProfile.soilType,
          ph: soilProfile.ph,
          n: soilProfile.nitrogen,
          p: soilProfile.phosphorus,
          k: soilProfile.potassium,
        },
        crop: cropId,
      }
    );

    return {
      schedule: (response.data.schedule || []).map((item: any) => ({
        type: item.type,
        quantity: item.quantity,
        unit: item.unit,
        timing: item.timing,
      })),
      organicAlternatives: (response.data.organic_alternatives || response.data.organicAlternatives || []).map((item: any) => ({
        type: item.type,
        quantity: item.quantity,
        unit: item.unit,
        timing: item.timing,
      })),
      soilAmendments: (response.data.soil_amendments || response.data.soilAmendments || []).map((item: any) => ({
        type: item.type,
        quantity: item.quantity,
        unit: item.unit,
        reason: item.reason,
      })),
    };
  } catch (err: unknown) {
    if (isAdvisoryEngineUnavailable(err)) {
      // Return mock data when Advisory Engine is unavailable
      return getMockFertilizerSchedule(cropId, soilProfile.ph, soilProfile.nitrogen, soilProfile.phosphorus, soilProfile.potassium);
    }
    throw err;
  }
}

// ─── Helpers ──────────────────────────────────────────────────────────────────

function isAdvisoryEngineUnavailable(err: unknown): boolean {
  if (!axios.isAxiosError(err)) return false;
  // 503 response from the engine
  if (err.response?.status === 503) return true;
  // Network error / ECONNREFUSED / no response (engine unreachable)
  if (!err.response) return true;
  return false;
}

/**
 * Determine the current agricultural season based on the current calendar month.
 * Indian seasons:
 * - kharif (Monsoon): July - October
 * - rabi (Winter): November - February
 * - zaid (Summer): March - June
 */
function getCurrentSeason(): string {
  const month = new Date().getMonth() + 1; // 1-12
  if (month >= 7 && month <= 10) {
    return 'kharif';
  } else if (month >= 11 || month <= 2) {
    return 'rabi';
  } else {
    return 'zaid';
  }
}

/**
 * Generate mock crop recommendations when the Advisory Engine is unavailable.
 * Provides realistic data based on soil conditions.
 */
function getMockCropRecommendations(
  ph: number,
  soilType: string | null,
  cropHistory: string[]
): CropRecommendation[] {
  // Common crops for Indian farmers
  const allCrops = [
    { name: 'Wheat', yieldRange: { min: 25, max: 35 }, waterRequirement: 'Medium (450-650mm)', estimatedInputCost: 15000, preferredPh: [6.0, 7.5] },
    { name: 'Rice', yieldRange: { min: 30, max: 45 }, waterRequirement: 'High (1200-1500mm)', estimatedInputCost: 18000, preferredPh: [5.5, 7.0] },
    { name: 'Maize', yieldRange: { min: 35, max: 50 }, waterRequirement: 'Medium (500-800mm)', estimatedInputCost: 12000, preferredPh: [5.8, 7.5] },
    { name: 'Cotton', yieldRange: { min: 15, max: 25 }, waterRequirement: 'Medium (600-800mm)', estimatedInputCost: 20000, preferredPh: [6.0, 8.0] },
    { name: 'Sugarcane', yieldRange: { min: 600, max: 900 }, waterRequirement: 'High (1500-2500mm)', estimatedInputCost: 35000, preferredPh: [6.5, 7.5] },
    { name: 'Groundnut', yieldRange: { min: 12, max: 20 }, waterRequirement: 'Low (400-600mm)', estimatedInputCost: 14000, preferredPh: [6.0, 6.5] },
    { name: 'Soybean', yieldRange: { min: 15, max: 25 }, waterRequirement: 'Medium (450-700mm)', estimatedInputCost: 13000, preferredPh: [6.0, 7.0] },
    { name: 'Mustard', yieldRange: { min: 8, max: 15 }, waterRequirement: 'Low (250-400mm)', estimatedInputCost: 8000, preferredPh: [6.5, 7.5] },
    { name: 'Chickpea', yieldRange: { min: 10, max: 18 }, waterRequirement: 'Low (300-500mm)', estimatedInputCost: 10000, preferredPh: [6.0, 7.5] },
    { name: 'Tomato', yieldRange: { min: 200, max: 350 }, waterRequirement: 'Medium (600-800mm)', estimatedInputCost: 25000, preferredPh: [6.0, 7.0] },
  ];

  // Score crops based on pH suitability
  const scoredCrops = allCrops.map(crop => {
    const [minPh, maxPh] = crop.preferredPh;
    let score = 10;
    
    // Reduce score if pH is out of optimal range
    if (ph < minPh) score -= (minPh - ph) * 2;
    if (ph > maxPh) score -= (ph - maxPh) * 2;
    score = Math.max(1, Math.min(10, score));

    // Avoid recommending the most recently grown crop at top
    if (cropHistory.length > 0 && crop.name.toLowerCase() === cropHistory[0].toLowerCase()) {
      score -= 3;
    }

    return { ...crop, score };
  });

  // Sort by score and return top 5
  const topCrops = scoredCrops.sort((a, b) => b.score - a.score).slice(0, 5);

  return topCrops.map(crop => ({
    name: crop.name,
    yieldRange: crop.yieldRange,
    waterRequirement: crop.waterRequirement,
    estimatedInputCost: crop.estimatedInputCost,
  }));
}

/**
 * Generate mock fertilizer schedule when the Advisory Engine is unavailable.
 * Provides realistic NPK recommendations based on actual soil deficiencies.
 */
function getMockFertilizerSchedule(
  cropId: string,
  ph: number | null | undefined,
  nitrogen: number | null | undefined,
  phosphorus: number | null | undefined,
  potassium: number | null | undefined
): FertilizerSchedule {
  // If soil values are missing, we can't provide accurate recommendations
  if (ph == null || nitrogen == null || phosphorus == null || potassium == null) {
    throw new AppError(
      'INCOMPLETE_SOIL_PROFILE',
      'Soil profile is missing nutrient data (N, P, K, pH). Please complete your soil profile.'
    );
  }

  const actualPh = ph;
  const actualN = nitrogen;
  const actualP = phosphorus;
  const actualK = potassium;

  const schedule: FertilizerSchedule['schedule'] = [];
  const organicAlternatives: FertilizerSchedule['organicAlternatives'] = [];
  const soilAmendments: FertilizerSchedule['soilAmendments'] = [];

  // NPK recommendations based on actual deficiency
  // Optimal ranges: N: 280-560 kg/ha, P: 10-25 kg/ha, K: 110-280 kg/ha
  
  // NITROGEN recommendations
  if (actualN < 200) {
    // Severely deficient
    schedule.push({
      type: 'Urea (46% N)',
      quantity: 200,
      unit: 'kg/acre',
      timing: 'Split: 40% at sowing, 30% at 30 days, 30% at 60 days',
    });
    organicAlternatives.push({
      type: 'Vermicompost',
      quantity: 600,
      unit: 'kg/acre',
      timing: 'Apply 3 weeks before sowing',
    });
  } else if (actualN < 280) {
    // Moderately deficient
    schedule.push({
      type: 'Urea (46% N)',
      quantity: 150,
      unit: 'kg/acre',
      timing: 'Split: 50% at sowing, 25% at 30 days, 25% at 60 days',
    });
    organicAlternatives.push({
      type: 'Vermicompost',
      quantity: 500,
      unit: 'kg/acre',
      timing: 'Apply 2 weeks before sowing',
    });
  } else if (actualN < 400) {
    // Slightly deficient
    schedule.push({
      type: 'Urea (46% N)',
      quantity: 100,
      unit: 'kg/acre',
      timing: 'Split: 60% at sowing, 40% at 30 days',
    });
    organicAlternatives.push({
      type: 'Farm Yard Manure (FYM)',
      quantity: 400,
      unit: 'kg/acre',
      timing: 'Apply 2 weeks before sowing',
    });
  } else if (actualN > 600) {
    // Excess nitrogen - skip or reduce
    schedule.push({
      type: 'Skip Nitrogen Fertilizer',
      quantity: 0,
      unit: 'kg/acre',
      timing: 'Soil has sufficient nitrogen. Monitor crop growth.',
    });
  }

  // PHOSPHORUS recommendations
  if (actualP < 5) {
    // Severely deficient
    schedule.push({
      type: 'Single Super Phosphate (SSP)',
      quantity: 150,
      unit: 'kg/acre',
      timing: 'Apply at sowing as basal dose',
    });
    organicAlternatives.push({
      type: 'Rock Phosphate',
      quantity: 120,
      unit: 'kg/acre',
      timing: 'Apply 3 weeks before sowing',
    });
  } else if (actualP < 10) {
    // Moderately deficient
    schedule.push({
      type: 'Single Super Phosphate (SSP)',
      quantity: 100,
      unit: 'kg/acre',
      timing: 'Apply at the time of sowing',
    });
    organicAlternatives.push({
      type: 'Rock Phosphate',
      quantity: 80,
      unit: 'kg/acre',
      timing: 'Apply 2 weeks before sowing',
    });
  } else if (actualP < 15) {
    // Slightly deficient
    schedule.push({
      type: 'DAP (Diammonium Phosphate)',
      quantity: 60,
      unit: 'kg/acre',
      timing: 'Apply at sowing',
    });
    organicAlternatives.push({
      type: 'Bone Meal',
      quantity: 50,
      unit: 'kg/acre',
      timing: 'Mix with soil before sowing',
    });
  } else if (actualP > 35) {
    // Excess phosphorus
    schedule.push({
      type: 'Skip Phosphorus Fertilizer',
      quantity: 0,
      unit: 'kg/acre',
      timing: 'Soil has sufficient phosphorus.',
    });
  }

  // POTASSIUM recommendations
  if (actualK < 80) {
    // Severely deficient
    schedule.push({
      type: 'Muriate of Potash (MOP)',
      quantity: 80,
      unit: 'kg/acre',
      timing: 'Split: 50% at sowing, 50% at first irrigation',
    });
    organicAlternatives.push({
      type: 'Wood Ash',
      quantity: 300,
      unit: 'kg/acre',
      timing: 'Mix with soil 2 weeks before sowing',
    });
  } else if (actualK < 110) {
    // Moderately deficient
    schedule.push({
      type: 'Muriate of Potash (MOP)',
      quantity: 50,
      unit: 'kg/acre',
      timing: 'Apply at sowing or first irrigation',
    });
    organicAlternatives.push({
      type: 'Wood Ash',
      quantity: 200,
      unit: 'kg/acre',
      timing: 'Mix with soil before sowing',
    });
  } else if (actualK < 160) {
    // Slightly deficient
    schedule.push({
      type: 'Potassium Sulphate',
      quantity: 30,
      unit: 'kg/acre',
      timing: 'Apply at sowing',
    });
  } else if (actualK > 350) {
    // Excess potassium
    schedule.push({
      type: 'Skip Potassium Fertilizer',
      quantity: 0,
      unit: 'kg/acre',
      timing: 'Soil has sufficient potassium.',
    });
  }

  // If all nutrients are in optimal range
  if (schedule.length === 0) {
    schedule.push({
      type: 'NPK Complex (19:19:19)',
      quantity: 60,
      unit: 'kg/acre',
      timing: 'Maintenance dose: Split 50% at sowing, 50% at 30 days',
    });
  }

  // pH amendments based on actual pH
  if (actualPh < 5.5) {
    // Very acidic
    soilAmendments.push({
      type: 'Agricultural Lime (Calcium Carbonate)',
      quantity: 300,
      unit: 'kg/acre',
      reason: `Soil pH (${actualPh.toFixed(1)}) is very acidic. Lime will raise pH to optimal range (6.0-7.5) over 3-6 months.`,
    });
  } else if (actualPh < 6.0) {
    // Moderately acidic
    soilAmendments.push({
      type: 'Agricultural Lime',
      quantity: 200,
      unit: 'kg/acre',
      reason: `Soil pH (${actualPh.toFixed(1)}) is acidic. Lime will raise pH to optimal range (6.0-7.5).`,
    });
  } else if (actualPh > 8.5) {
    // Very alkaline
    soilAmendments.push({
      type: 'Gypsum (Calcium Sulphate)',
      quantity: 300,
      unit: 'kg/acre',
      reason: `Soil pH (${actualPh.toFixed(1)}) is very alkaline. Gypsum will help lower pH to optimal range (6.0-7.5).`,
    });
  } else if (actualPh > 8.0) {
    // Moderately alkaline
    soilAmendments.push({
      type: 'Gypsum',
      quantity: 250,
      unit: 'kg/acre',
      reason: `Soil pH (${actualPh.toFixed(1)}) is alkaline. Gypsum will help lower pH to optimal range (6.0-7.5).`,
    });
  }

  // Micronutrients (always recommended for Indian soils)
  schedule.push({
    type: 'Zinc Sulphate',
    quantity: 10,
    unit: 'kg/acre',
    timing: 'Apply once at sowing (micronutrient)',
  });

  // Add iron if pH is too high (alkaline soils lock iron)
  if (actualPh > 8.0) {
    schedule.push({
      type: 'Ferrous Sulphate',
      quantity: 12,
      unit: 'kg/acre',
      timing: 'Apply at sowing (alkaline soil correction)',
    });
  }

  return {
    schedule,
    organicAlternatives,
    soilAmendments,
  };
}
